File: data_structures/trees/binary_tree/typescript/binaryTreeIsSame.ts
import BinaryTreeNode from '@ds/trees/binary_tree/typescript/BinaryTreeNode';
import Stack from '@ds/stack/typescript/Stack';

/**
 * Function that validates recursively if two binary trees are equal
 * @param p
 * @param q
 */
function binaryTreeIsSameRecursive<T>(p: BinaryTreeNode<T>, q: BinaryTreeNode<T>): boolean {
    if (p && q) {
        const left = binaryTreeIsSameRecursive(p.left, q.left);
        const right = binaryTreeIsSameRecursive(p.right, q.right);

        return p.val === q.val && left && right;
    } else {
        return p === q;
    }
}

function binaryTreeIsSameStack<T>(p: BinaryTreeNode<T>, q: BinaryTreeNode<T>): boolean {
    const stack = new Stack<[BinaryTreeNode<T>, BinaryTreeNode<T>]>();
    stack.push([p, q]);

    while (!stack.isEmpty) {
        const [node1, node2] = stack.pop();

        if (node1 && node2 && node1.val === node2.val) {
            stack.push([node1.right, node2.right]);
            stack.push([node1.left, node2.left]);
        } else if (!node1 && !node2) {
            continue;
        } else {
            return false;
        }
    }

    return true;
}

export { binaryTreeIsSameRecursive, binaryTreeIsSameStack };

File: data_structures/trees/binary_tree/python/binary_tree_is_same.py
from typing import TypeVar, Tuple
from data_structures.trees.binary_tree.python.binary_tree_node import BinaryTreeNode
from data_structures.stack.pyhton.stack import Stack

T = TypeVar('T')


def binary_tree_is_same_recursive(p: BinaryTreeNode[T], q: BinaryTreeNode[T]) -> bool:
    if p and q:
        left = binary_tree_is_same_recursive(p.left, q.left)
        right = binary_tree_is_same_recursive(p.right, q.right)

        return p.val == q.val and left and right
    else:
        return p == q


def binary_tree_is_same_stack(p: BinaryTreeNode[T], q: BinaryTreeNode[T]) -> bool:
    StackType = Tuple[BinaryTreeNode[T], BinaryTreeNode[T]]
    stack = Stack[StackType]()
    stack.push([p, q])

    while not stack.is_empty:
        popped: StackType = stack.pop()
        n1, n2 = popped

        if n1 and n2 and n1.val == n2.val:
            stack.push((n1.right, n2.right))
            stack.push((n1.left, n2.left))
        elif not n1 and not n2:
            continue
        else:
            return False

    return True

